### PR TITLE
Extended Api Dataset

### PR DESCRIPTION
**Pull request by AzanWasem**
📌 Description of the Code

This notebook demonstrates how to use the TensorFlow Dataset API with the MNIST dataset to train a simple Convolutional Neural Network (CNN).

1. Introduction

The code uses TensorFlow’s Dataset API, which provides an efficient data pipeline for feeding data into a model.

It loads MNIST digit images (handwritten numbers 0–9) and prepares them for training.

2. Loading the Data

MNIST dataset is imported in NumPy format.

Data is stored in /tmp/data/ and extracted automatically.

3. Parameters Setup

Learning rate = 0.01

Training steps = 1000

Batch size = 128

Input features = 784 (28×28 pixels)

Output classes = 10 (digits 0–9)

Dropout rate = 0.75

4. Dataset Pipeline

Creates a TensorFlow Dataset from MNIST images and labels.

Repeats data indefinitely, batches it into size 128, and prefetches for faster performance.

An iterator is initialized to provide batches during training.

5. Convolutional Neural Network (CNN)

Reshape Input: Images reshaped into [batch_size, 28, 28, 1].

Conv Layer 1: 32 filters, kernel size 5, ReLU activation + MaxPooling.

Conv Layer 2: 64 filters, kernel size 3, ReLU activation + MaxPooling.
Flatten Layer: Converts 2D output into 1D vector.
Fully Connected Layer: 1024 neurons + Dropout.
Output Layer: Dense layer → logits for 10 classes.
Uses softmax for predictions in testing.

6. Training Setup
Loss Function: softmax_cross_entropy_with_logits.
Optimizer: Adam with learning rate 0.01.
Metrics: Accuracy is calculated by comparing predicted vs. true labels.

7. Training Loop
Initializes all variables.
Runs optimization for 1000 steps.
Every 100 steps, prints minibatch loss and training accuracy.
Ends with Optimization Finished!.

8. Output Example
Training accuracy improves from around 7% (random guessing) to 96%, showing that the CNN successfully learns to classify handwritten digits.

👉 In short:
This code shows how to use TensorFlow’s Dataset API to build an efficient input pipeline and train a CNN on MNIST with dropout regularization and batching.

### CODE REVIEW

**Review by AzanWasem — COMMENTED** [review]
📌 Description of the Code

This notebook demonstrates how to use the TensorFlow Dataset API with the MNIST dataset to train a simple Convolutional Neural Network (CNN).

1. Introduction

The code uses TensorFlow’s Dataset API, which provides an efficient data pipeline for feeding data into a model.

It loads MNIST digit images (handwritten numbers 0–9) and prepares them for training.

2. Loading the Data

MNIST dataset is imported in NumPy format.

Data is stored in /tmp/data/ and extracted automatically.

3. Parameters Setup

Learning rate = 0.01

Training steps = 1000

Batch size = 128

Input features = 784 (28×28 pixels)

Output classes = 10 (digits 0–9)

Dropout rate = 0.75

4. Dataset Pipeline

Creates a TensorFlow Dataset from MNIST images and labels.

Repeats data indefinitely, batches it into size 128, and prefetches for faster performance.

An iterator is initialized to provide batches during training.

5. Convolutional Neural Network (CNN)

Reshape Input: Images reshaped into [batch_size, 28, 28, 1].

Conv Layer 1: 32 filters, kernel size 5, ReLU activation + MaxPooling.

Conv Layer 2: 64 filters, kernel size 3, ReLU activation + MaxPooling.

Flatten Layer: Converts 2D output into 1D vector.

Fully Connected Layer: 1024 neurons + Dropout.

Output Layer: Dense layer → logits for 10 classes.

Uses softmax for predictions in testing.

6. Training Setup

Loss Function: softmax_cross_entropy_with_logits.

Optimizer: Adam with learning rate 0.01.

Metrics: Accuracy is calculated by comparing predicted vs. true labels.

7. Training Loop

Initializes all variables.

Runs optimization for 1000 steps.

Every 100 steps, prints minibatch loss and training accuracy.

Ends with Optimization Finished!.

8. Output Example

Training accuracy improves from around 7% (random guessing) to 96%, showing that the CNN successfully learns to classify handwritten digits.

👉 In short:
This code shows how to use TensorFlow’s Dataset API to build an efficient input pipeline and train a CNN on MNIST with dropout regularization and batching.

**Comment by AzanWasem** [comment]
AzanWasem-patch-1

**Comment by AzanWasem** [comment]
git pull origin main
